### PR TITLE
Top-Level Project Generation Change

### PR DESCRIPTION
**Pull request by andreiborodin**
This change updates how we perform top-level project generation, and a bit of other cleanup.

### CODE REVIEW

File: Source/MSBuildTools.Unity/Packages/com.microsoft.msbuildforunity/Editor/ProjectGenerator/Scripts/MSBuildTools.cs [hunk 1]
```diff
@@ -237,7 +233,7 @@ public static void GenerateSDKProjects()
 
         public static void RegenerateSDKProjects()
         {
-            RegenerateEverything(reparseUnityData: true);
+            RegenerateEverything(unityProjectInfo = new UnityProjectInfo(SupportedBuildTargets, Config, true));
```
Comment (chrisfromwork) [inline Source/MSBuildTools.Unity/Packages/com.microsoft.msbuildforunity/Editor/ProjectGenerator/Scripts/MSBuildTools.cs]: >true [](start = 104, length = 4)

nit: specify the argument name here for the true.

File: Source/MSBuildTools.Unity/Packages/com.microsoft.msbuildforunity/Editor/ProjectGenerator/Scripts/MSBuildUnityProjectExporter.cs [hunk 2]
```diff
@@ -75,6 +78,38 @@ public static void ExportCommonPropsFile(IUnityProjectExporter exporter, Compila
             propsExporter.Write();
         }
 
+        public static void ExportTopLevelDependenciesProject(IUnityProjectExporter exporter, MSBuildToolsConfig config, DirectoryInfo generatedProjectFolder, UnityProjectInfo unityProjectInfo)
+        {
```
Comment (chrisfromwork) [inline Source/MSBuildTools.Unity/Packages/com.microsoft.msbuildforunity/Editor/ProjectGenerator/Scripts/MSBuildUnityProjectExporter.cs]: nit: add summary

File: Source/MSBuildTools.Unity/Packages/com.microsoft.msbuildforunity/Editor/ProjectGenerator/Scripts/UnityProjectInfo.cs [hunk 3]
```diff
@@ -104,32 +118,59 @@ public UnityProjectInfo(Dictionary<BuildTarget, string> supportedBuildTargets, M
...
 
-            CurrentPlayerPlatform = AvailablePlatforms.First(t => t.BuildTarget == EditorUserBuildSettings.activeBuildTarget);
+            if (performCompleteParse)
+            {
+                RefreshProjects(config);
+            }
         }
 
         public void Dispose()
         {
```
Comment (chrisfromwork) [inline Source/MSBuildTools.Unity/Packages/com.microsoft.msbuildforunity/Editor/ProjectGenerator/Scripts/UnityProjectInfo.cs]: nit: add summary

File: Source/MSBuildTools.Unity/Packages/com.microsoft.msbuildforunity/Editor/ProjectGenerator/Scripts/UnityProjectInfo.cs [hunk 4]
```diff
@@ -104,32 +118,59 @@ public UnityProjectInfo(Dictionary<BuildTarget, string> supportedBuildTargets, M
...
+                WinMDs = new ReadOnlyCollection<WinMDInfo>(winmds);
+
+                foreach (PluginAssemblyInfo plugin in Plugins)
+                {
+                    if (plugin.Type == PluginType.Native)
+                    {
+                        // Logging will be re-enabled with robust update holistically across MSB4U: https://github.com/microsoft/MSBuildForUnity/issues/75
+                        // Debug.Log($"Native plugin {plugin.ReferencePath.AbsolutePath} not yet supported for MSBuild project.");
+                    }
                 }
```
Comment (chrisfromwork) [inline Source/MSBuildTools.Unity/Packages/com.microsoft.msbuildforunity/Editor/ProjectGenerator/Scripts/UnityProjectInfo.cs]: nit: remove this foreach if its a noop. Or comment it all out

File: Source/MSBuildTools.Unity/Packages/com.microsoft.msbuildforunity/Editor/ProjectGenerator/Scripts/Exporters/ITopLevelDependenciesProjectExporter.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        /// </summary>
+        public Uri ReferencePath { get; set; }
+
+        /// <summary>
+        /// Condition for this reference.
+        /// </summary>
+        public string Condition { get; set; }
+
+        public override int GetHashCode()
+        {
```
Comment (chrisfromwork) [inline Source/MSBuildTools.Unity/Packages/com.microsoft.msbuildforunity/Editor/ProjectGenerator/Scripts/Exporters/ITopLevelDependenciesProjectExporter.cs]: nit: add inheritdoc for these overrides

**Comment by chrisfromwork** [comment]
How were these changes tested?

File: Source/MSBuildTools.Unity/Packages/com.microsoft.msbuildforunity/Editor/ProjectGenerator/Scripts/MSBuildUnityProjectExporter.cs (hunk 2 above)
Comment (andreiborodin) [reply Source/MSBuildTools.Unity/Packages/com.microsoft.msbuildforunity/Editor/ProjectGenerator/Scripts/MSBuildUnityProjectExporter.cs]: Leaving for now, as this class is temporary so I can split off the logic and then adapt it in the appropriate place.

---
In reply to: [369676752](https://github.com/microsoft/MSBuildForUnity/pull/111#discussion_r369676752) [](ancestors = 369676752)

**Comment by andreiborodin** [comment]
Tested this with all three samples, and actually the work you been doing in SpectatorView.Unity

---
In reply to: [577283613](https://github.com/microsoft/MSBuildForUnity/pull/111#issuecomment-577283613) [](ancestors = 577283613)

**Pull request merged**